### DISCUSSION TITLE
How describe model for Record<string, string> TS type?

### DISCUSSION THREAD

**Discussion by KyryloSWAN** [mobxjs/mobx-state-tree#2003] [Q&A] [answered]
I have the following data for which I need to create a model:
```
{
  "some": [
    {"a": "b"},
    {"c": "d"},
    {"e": "f"}
  ]
}
```
The key of object in array item can be any string.
I can use `Array<Record<string, string>>`  to describe my data as TypeScript type.
But I can't figure out how I can describe the mst model for my data.

**Comment by coolsoftwaretyler** [comment, marked as answer]
Hey @KyryloSWAN - here's how I might model this out to start

```js
const TopLevelObject = types
  .model("TopLevelObject", {
    some: types.array(types.frozen),
  });
```

It seems like the keys in each of your records are dynamic, and might not be suitable for "typing" in advance. In this case, the [frozen type](https://mobx-state-tree.js.org/API/#frozen) allows you to work with these objects, since they're serializable.

I do this from time to time when it comes with new third party APIs, where I'm not yet sure what data I'm going to really want or need to model out from them. We sometimes store JSON blobs as `frozen` types, until we can determine a better way to model the data that we expect to use from it.

**Reply by KyryloSWAN** [reply]
Thank you, @coolsoftwaretyler!
frozen works!

**Reply by KyryloSWAN** [reply]
I'm trying to use `types.frozen<Record<string, string>>() ` to decribe model for `Array<Record<string, string>> | undefined` type:
```
export const ProfileStore = types
  .model({
    autoExchangeDestinations: types.maybe(
      types.array(types.frozen<Record<string, string>>())
    ),
})
  .actions((self) => {
    return {
      getUserProfileById: flow(function* getUserProfile(id: string) {
...
          self.autoExchangeDestinations =
                    fetchedArray === null
                      ? undefined
                      : [...fetchedArray];
        }
      }),
    };
  });
```
where `fetchedArray` has `Record<string, string>[] | null` type.
And got error TS2322:
```
error TS2322: Type 'Record<string, string>[] | undefined' is not assignable to type '(IMSTArray<IType<Record<string, string>, Record<string, string>, Record<string, string>>> & IStateTreeNode<IMaybe<IArrayType<IType<Record<string, string>, Record<...>, Record<...>>>>>) | undefined'.
  Type 'Record<string, string>[]' is not assignable to type 'IMSTArray<IType<Record<string, string>, Record<string, string>, Record<string, string>>> & IStateTreeNode<IMaybe<IArrayType<IType<Record<string, string>, Record<...>, Record<...>>>>>'.
    Type 'Record<string, string>[]' is missing the following properties from type 'IMSTArray<IType<Record<string, string>, Record<string, string>, Record<string, string>>>': spliceWithArray, clear, replace, remove, toJSON

114                   self.autoExchangeDestinations =
                      ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
What I'm doing wrong?

**Reply by coolsoftwaretyler** [reply]
Ah yes, that looks like a known problem with TS, MST, and arrays. Try the solution listed here: https://github.com/mobxjs/mobx-state-tree/issues/957#issuecomment-410075001

Let me know if that helps or not.

**Reply by KyryloSWAN** [reply]
It helps!
```
                  self.autoExchangeDestinations =
                    autoExchangeDestinations === null
                      ? undefined
                      : cast([...autoExchangeDestinations]);
```
works with flying colors!
Thank you for your help @coolsoftwaretyler
